### PR TITLE
#164396754 Ft Add Pagination

### PR DESCRIPTION
**Pull request by Nerldy**
#### What does this PR do?

- Returns pagination data

#### Description of Task to be completed?
- This will provide a smaller chunk of a report to the user, help him/her to read the report page easier and also include parameter based filtering.

#### How should this be manually tested?
1. Clone the project and run it
2. On an app like Postman, send url in this format `/api/v1/automations?page={int}&limit={int}`; 
**limit** being the number of items you expect the pagination to return, and **page** being the page number: **### example below**

3lwmpgdfsb

#### What are the relevant pivotal tracker stories?

[#164396754](https://www.pivotaltracker.com/n/projects/2197045/stories/164396754)

### CODE REVIEW

File: test/endpoints/automations.test.js [hunk 1]
```diff
@@ -20,4 +20,17 @@ describe('Tests for automation endpoints\n', () => {
         done();
       });
   });
+
+  it('Should return a 500 server error', (done) => {
+    chai
+      .request(app)
+      .get('/api/v1/automations')
```
Comment (Veraclins) [inline test/endpoints/automations.test.js]: Great job @Nerldy. I am curious, why is this an error? And why a 500 error?

File: server/controllers/automationController.js [hunk 2]
```diff
@@ -18,11 +19,55 @@ export default class AutomationController {
...
-      status: 'success',
-      message: 'Successfully fetched automations',
-      data: formatAutomationResponse(data),
-    }));
+    const limit = +req.query.limit;
+    let offset = 0;
+    let prevPage;
+    let nextPage;
+
+    // include[2].where = {
```
Comment (asamisellah) [inline server/controllers/automationController.js]: Kindly remove the commented code to make the code clean

File: server/controllers/automationController.js [hunk 3]
```diff
@@ -18,11 +19,55 @@ export default class AutomationController {
...
+              currentPage: page,
+              numberOfPages,
+              dataCount: data.count,
+              nextPage,
+              prevPage,
+            },
+          }));
+      })
+      .catch((error) => {
+        console.log(error);
```
Comment (asamisellah) [inline server/controllers/automationController.js]: Kindly remove the console statement

Comment (Nerldy) [reply server/controllers/automationController.js]: done.

File: test/endpoints/automations.test.js (hunk 1 above)
Comment (Nerldy) [reply test/endpoints/automations.test.js]: Thanks. It was to catch an error if the url didn't have the correct queries but thanks to @valmayaki suggestion I've gone instead with a default value and removed the catch function.

File: server/controllers/automationController.js [hunk 4]
```diff
@@ -18,11 +19,55 @@ export default class AutomationController {
    * @param {object} res REST Response object
    * @returns {object} Response containing status message and automation data
    */
+
   static getAutomations(req, res) {
```
Comment (valmayaki) [inline server/controllers/automationController.js]: to fix the code climate issue with code to long try this:
```javascript
static async getAutomations(req, res) {
    const modelOptions = {
      include,
      order,
    };
    const response = {
      status: 'success',
      message: 'Successfully fetched automations',
    };
    let countData;
    // if url doesn't contain a query objects send back all the data
    if ((Object.entries(req.query).length !== 0 && req.query.constructor === Object) || req.query.limit !== -1) {
      countData = await automation.findAndCountAll();
      response.pagination = {};
      response.pagination.page = parseInt(req.query.page, 10) || 1; // current page number
      const limit = parseInt(req.query.limit, 10) || 10;
      const offset = limit * (response.pagination.page - 1);
      response.pagination.numberOfPages = Math.ceil(countData.count / limit); // all pages count

      // check if number of pages is less than the current page number to show next page number
      if (response.pagination.page < response.pagination.numberOfPages) {
        response.pagination.nextPage = response.pagination.page + 1;
      }
      // show previous page number if page is greater than 1
      if (response.pagination.page > 1) {
        response.pagination.prevPage = response.pagination.page - 1;
      }
      modelOptions.offset = offset;
      modelOptions.limit = limit;
    }
    const allData = await automation.findAll({ ...modelOptions });
    response.data = formatAutomationResponse(allData);
    res.status(200).json(response);
  }
```
you can still reduce it further by separating some part into private functions

Comment (Nerldy) [reply server/controllers/automationController.js]: Thanks. Working on it.

**Pull request merged**